src/codelabs/async-await/index.md

```diff
@@ -0,0 +1,473 @@
+main() [!async!] {
+  print('Fetching user order...');
+  print([!await!] createOrderMessage());
+}
+
+// "Your order is: Large Latte"
+{% endprettify %}
+    </div>
+  </div>
+</div>
+
+The asynchronous example is different in three ways:
+  * The return type for `createOrderMessage()` changes from `String` to `Future<String>`.
+  * The **`async`** keyword appears before the function bodies for `createOrderMessage()` and the `main()` function.
+  * The **`await`** keyword appears before calling the asynchronous functions `getUserOrder()` and `createOrderMessage()`.
+
+{{ site.alert.secondary }}
+  **Key terms:**
+* **async**: You can use the `async` keyword before a function's body to mark it as
+asynchronous.
+* **async function**:  An `async` function is a function labeled with the `async` keyword. 
+* **await**: You can use the `await` keyword to get the completed result of an
+asynchronous expression. You can only use the `await` keyword within an `async` function. 
+{{ site.alert.end }}
+
+## Execution flow with async and await
+
+As of Dart 2, functions marked as `async` run synchronously until the first
+`await` keyword. This means that within an `async` function body, all
+synchronous code before the first `await` keyword executes immediately. 
+
+### Example: Execution within async functions
+Run the following example to see how execution proceeds within an `async` 
+function body. What do you think the output will be?
+
+[//]: https://gist.github.com/d7abfdea1ae5596e96c7c0203d975dba
+<iframe 
+  src="https://dartpad.dartlang.org/experimental/embed-new-dart.html?id=d7abfdea1ae5596e96c7c0203d975dba"
+  style="border: 1px solid lightgrey; margin-top: 10px; margin-bottom: 40px"
+  frameborder="no"
+  height="600"
+  width="100%">
+</iframe>
+
+After running the code in the preceding example, try reversing line 4 and line 5:
+
+{% prettify dart %}
+  var order = await getUserOrder();
+  print('Awaiting user order...');
+{% endprettify %}
+
+Notice that the timing of the output shifts, now that the `print('Awaiting user order')`
+statement appears after the first `await` keyword in `createOrderMessage()`.
+
+### Exercise: Practice using async and await
+
+The following exercise is a failing unit test that contains partially completed
+code snippets. Your task is to complete the exercise by writing code to make the 
+tests pass. 
+You don't need to implement `main()`.
+
+To simulate asynchronous operations, your code will call the following functions, which are 
+provided for you:
+
+<!-- dart annotation format borrowed from  
+https://api.dartlang.org/stable/2.4.0/dart-developer/dart-developer-library.html -->
+<!-- See e.g. getCurrentTag() → UserTag -->
+
+{% prettify dart %}
+getRole() → Future<String> 
+// Returns a short description of the user's role
+
+getLoginAmount() → Future<int> 
+// Returns the number of times a user has logged in
+{% endprettify %}
+#### Part 1: `reportUserRole()`
+
+Add logic to the `reportUserRole()` function:
```

Comment:
"Add logic" doesn't sound right, and it could be easy to misread the following bullets as being already true, rather than goals. Maybe:

Add code to the `reportUserRole` function so that it does the following: